### PR TITLE
First pass at obstest script. (#73)

### PR DESCRIPTION
**Pull request by broomweed**
This script reads in `.obstest` files which are specified like this:
```
server1: file1.obs
server2: file2.obs
client: client.obs
***
expected output
```
It locates these files relative to the location of the obstest script, then compiles each of the files, compiling the `client` contract with the `--build-client` flag. It forks off processes to run each non-client 
process on a different port, then passes each one to the client (on the command line, for now), and checks that the output of the client matches what's specified in the file using `diff`.

It's written in Perl. I think it was easier than trying to parse text files using bash...

This is just a first pass at this script -- it might want to support changing the port number, and it doesn't yet support running these things on the actual blockchain. Also, the client doesn't actually handle receiving more than one server yet.

(Also, it's not very good at figuring out whether compilation failed. I'm not sure if that's SBT or the Obsidian compiler's fault, but if we could get a success/failure return code, that would be great.)

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -6,6 +6,8 @@ scalaVersion := "2.12.6"
 
 libraryDependencies += "org.scala-lang.modules" %% "scala-parser-combinators" % "1.0.6"
 
+trapExit := false
```
Comment (mcoblenz) [inline build.sbt]: What is this for? Please add a comment.

File: obstest.pl [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    # Parse test file.
+    my ($namesref, $serversref, $propsref) = parse_obstest($filename);
+    # (unpack returned data structures into real variables. Perl)
+    my @servernames = @$namesref;
+    my %servers = %$serversref;
+    my %props = %$propsref;
+
+    die "Must specify a file to designate as the client.\n" unless exists $servers{client};
+
+    # Since we process client differently, we split it into its own thing.
```
Comment (mcoblenz) [inline obstest.pl]: thing?

**Comment by mcoblenz** [comment]
Here is what I get when I try:

````
GS14461:Obsidian mcoblenz 08:51$ ./obstest.pl ./resources/tests/TestIntContainer.obstest
./resources/tests/TestIntContainer.obstest
    server => ./resources/tests/IntContainer.obs
    client => ./resources/tests/IntContainerClient.obs
sbt "run --verbose --build-client --dump-debug obs_output --output-path obs_output ./resources/tests/IntContainerClient.obs"
Test ./resources/tests/TestIntContainer.obstest failed: sbt did not say 'success':
[info] Loading settings from assembly.sbt ...
[info] Loading project definition from /Users/mcoblenz/code/Obsidian/project
[info] Loading settings from build.sbt ...
[info] Set current project to Obsidian (in build file:/Users/mcoblenz/code/Obsidian/)
[info] Compiling 17 Scala sources to /Users/mcoblenz/code/Obsidian/target/scala-2.12/classes ...
[info] Done compiling.
[info] Packaging /Users/mcoblenz/code/Obsidian/target/scala-2.12/obsidian_2.12-1.0.jar ...
[error] java.util.zip.ZipException: duplicate entry: META-INF/MANIFEST.MF
[error] 	at java.util.zip.ZipOutputStream.putNextEntry(ZipOutputStream.java:232)
[error] 	at java.util.jar.JarOutputStream.putNextEntry(JarOutputStream.java:109)
[error] 	at sbt.io.IO$.addFileEntry$1(IO.scala:539)
[error] 	at sbt.io.IO$.$anonfun$writeZip$3(IO.scala:548)
[error] 	at sbt.io.IO$.$anonfun$writeZip$3$adapted(IO.scala:548)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:929)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:929)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:71)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
[error] 	at sbt.io.IO$.writeZip(IO.scala:548)
[error] 	at sbt.io.IO$.$anonfun$archive$1(IO.scala:502)
[error] 	at sbt.io.IO$.$anonfun$archive$1$adapted(IO.scala:499)
[error] 	at sbt.io.IO$.$anonfun$withZipOutput$1(IO.scala:584)
[error] 	at sbt.io.IO$.$anonfun$withZipOutput$1$adapted(IO.scala:573)
[error] 	at sbt.io.Using.apply(Using.scala:22)
[error] 	at sbt.io.IO$.withZipOutput(IO.scala:573)
[error] 	at sbt.io.IO$.archive(IO.scala:499)
[error] 	at sbt.io.IO$.jar(IO.scala:478)
[error] 	at sbt.Package$.makeJar(Package.scala:117)
[error] 	at sbt.Package$.$anonfun$apply$4(Package.scala:72)
[error] 	at sbt.Package$.$anonfun$apply$4$adapted(Package.scala:70)
[error] 	at sbt.util.Tracked$.$anonfun$inputChanged$1(Tracked.scala:149)
[error] 	at sbt.Package$.apply(Package.scala:80)
[error] 	at sbt.Defaults$.$anonfun$packageTask$1(Defaults.scala:1073)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:42)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:64)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:257)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.Execute.work(Execute.scala:266)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:257)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:167)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:32)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[error] 	at java.lang.Thread.run(Thread.java:745)
[error] (compile:packageBin) java.util.zip.ZipException: duplicate entry: META-INF/MANIFEST.MF
````

**Comment by mcoblenz** [comment]
I bet I need to apply this to the branch from #122, don't I?

**Comment by mcoblenz** [comment]
I'm still getting the same error as above ((compile:packageBin) java.util.zip.ZipException: duplicate entry: META-INF/MANIFEST.MF)

File: obstest.pl [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+    print $cmd, "\n";
+    my $output = `$cmd`;
+
+    if ($verbose) {
+        print "Client output:\n", $output;
+    }
+
+    kill 'HUP', @pids;
+    sleep 1; # wait for processes to shut down
```
Comment (mcoblenz) [inline obstest.pl]: shouldn't this be a wait call rather than a sleep?

**Comment by broomweed** [comment]
For posterity: the file that had to be removed/moved in order to fix the SBT error was `target/scala-2.12/classes/META-INF/MANIFEST.MF`, in case we run into this issue in the future.

**Pull request merged**